File: matlab/examples/ex_hdr_images.m
% This example shows how to run FovVideoVDP on HDR images.

if ~exist( 'fovvdp', 'file' )
    addpath( fullfile( pwd, '..') );
end

I_ref = hdrread( 'nancy_church.hdr' );

L_peak = 4000; % Peak luminance of an HDR display

% HDR images are often given in relative photometric units. They MUST be
% mapped to absolute amount of light emitted from the display. For that, 
% we map the peak value in the image to the peak value of the display,
% then we increase the brightness by 2 stops (*4):
I_ref = I_ref/max(I_ref(:)) * L_peak * 4;

% Add Gaussian noise of 20% contrast
I_test_noise = I_ref + I_ref.*randn(size(I_ref))*0.3;

I_test_blur = imgaussfilt( I_ref, 2 );

% We use geometry of SDR 4k 30" display, but ignore its photometric
% properties and instead tell that we pass absolute colorimetric values. 
% Note that many HDR images are in rec709 color space, so no need to
% specify rec2020. 
[Q_JOD_noise, diff_map_noise] = fvvdp( I_test_noise, I_ref, 'display_name', 'standard_hdr', 'display_photometry', 'absolute', 'heatmap', 'threshold' );
[Q_JOD_blur, diff_map_blur] = fvvdp( I_test_blur, I_ref, 'display_name', 'standard_hdr', 'display_photometry', 'absolute', 'heatmap', 'threshold' );

clf
subplot( 1, 2, 1 );
imshow( diff_map_noise );
title( sprintf('Quality: %g JOD\n', Q_JOD_noise ) );

subplot( 1, 2, 2 );
imshow( diff_map_blur );
title( sprintf('Quality: %g JOD\n', Q_JOD_blur ) );

File: pytorch_examples/ex_hdr_images.py
import os
import numpy as np
import torch
import matplotlib.pyplot as plt
import ex_utils as utils

import pyfvvdp

I_ref = pyfvvdp.load_image_as_array(os.path.join('example_media', 'nancy_church.hdr'))

noise_fname = os.path.join('example_media', 'wavy_facade_noise.png')
L_peak = 4000   # Peak luminance of an HDR display

# HDR images are often given in relative photometric units. They MUST be
# mapped to absolute amount of light emitted from the display. For that, 
# we map the peak value in the image to the peak value of the display,
# then we increase the brightness by 2 stops (*4). This is an arbitrary 
# choise (of colour grading/tone mapping), and different mapping could be used. 
I_ref = I_ref/I_ref.max() * L_peak * 4

# Add Gaussian noise of 20% contrast
I_test_noise = (I_ref + I_ref*np.random.randn(*I_ref.shape)*0.3).astype(I_ref.dtype)

I_test_blur = utils.imgaussblur(I_ref, 2)

# We use geometry of SDR 4k 30" display, but ignore its photometric
# properties and instead tell that we pass absolute linear colorimetric values. 
# (the display name is 'standard_hdr_linear')
# Note that many HDR images are in rec709 color space, so no need to
# specify rec2020. 
disp_photo = pyfvvdp.fvvdp_display_photo_absolute(L_peak)
fv = pyfvvdp.fvvdp(display_name='standard_hdr_linear', display_photometry=disp_photo, heatmap='threshold')

# predict() method can handle numpy ndarrays or PyTorch tensors. The data
# type should be float32, int16 or uint8.
# Channels can be in any order, but the order must be specified as a dim_order parameter. 
# Here the dimensions are (Height,Width,Colour)
Q_JOD_noise, stats_noise = fv.predict( I_test_noise, I_ref, dim_order="HWC" )
noise_str = f'Noise - Quality: {Q_JOD_noise:.3f} JOD'
print( noise_str )

Q_JOD_blur, stats_blur = fv.predict( I_test_blur, I_ref, dim_order="HWC" )
blur_str = f'Blur - Quality: {Q_JOD_blur:.3f} JOD'
print( blur_str )

f, axs = plt.subplots(1, 2)
axs[0].imshow( stats_noise['heatmap'][0,:,0,:,:].permute([1,2,0]).to(torch.float32).cpu().numpy() )
axs[0].set_xticks([])
axs[0].set_yticks([])
axs[0].set_title(noise_str)
axs[1].imshow( stats_blur['heatmap'][0,:,0,:,:].permute([1,2,0]).to(torch.float32).cpu().numpy() )
axs[1].set_xticks([])
axs[1].set_yticks([])
axs[1].set_title(blur_str)

f.show()
plt.waitforbuttonpress()
